UI for the Classroom Admin Page

Nik-09 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

The PR creates the UI for the Classroom Admin Page so that the curriculum admin can edit/create/delete a classroom.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct




https://user-images.githubusercontent.com/57531197/189813079-b46b199b-a33b-4814-ae9c-8f0f07651318.mp4








## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Nik-09 committed:
Adds diagnostic test model.

Nik-09 committed:
Created Classroom model.

Nik-09 committed:
Adds backend test into the shard.

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Merge branch 'classroom_model' of https://github.com/Nik-09/oppia into classroom_model

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Adds backend coverage tests

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Fixes backend test.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
NMerge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Adds domain object for classroom model.

Nik-09 committed:
Fixes lint issue.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
fixes lint issue

Nik-09 committed:
Merge branch 'classroom_model' of https://github.com/Nik-09/oppia into domain_object_for_classroom_model

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into domain_object_for_classroom_model

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Adds more backend unit-tests

Nik-09 committed:
Adds handler class for the classroom admin page

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into domain_object_for_classroom_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into domain_object_for_classroom_model

Nik-09 committed:
Merge branch 'domain_object_for_classroom_model' of https://github.com/Nik-09/oppia into controller_layer_addition_for_classroom

Nik-09 committed:
Adds classroom config services file

Nik-09 committed:
Adds test case structure

Nik-09 committed:
Adds test for the classroom config service

Nik-09 committed:
Adds backend unit tests for classroom admin handlers.

Nik-09 committed:
Fixes mypy, linter, backend unit tests.

Nik-09 committed:
Fixes lint issues.

Nik-09 committed:
Fixes merge conflicts

Nik-09 committed:
Adds backend unit test.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into controller_and_service_changes_for_classroom

Nik-09 committed:
Merge branch 'controller_and_service_changes_for_classroom' of https://github.com/Nik-09/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Adds backend api for the classroom admin.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into controller_and_service_changes_for_classroom

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Merge branch 'controller_and_service_changes_for_classroom' of https://github.com/Nik-09/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Fixes lint issues.

Nik-09 committed:
Created diecrtory srructure for the classroom admin UI.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Adds admin feature to generate dummy classroom data

Nik-09 committed:
Add extra newlines for readability

Nik-09 committed:
Makes UI more interactive.

Nik-09 committed:
Adds modals for confirmation in various events

Nik-09 committed:
Merge branch 'develop' into backend_api_changes_for_classroom_admin

Nik-09 committed:
Refactored code for increasing readability.

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Adds copyright comments in the starting of newly created files.

Nik-09 committed:
Adds textarea instead of input

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Merge branch 'backend_api_changes_for_classroom_admin' of https://github.com/Nik-09/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Merge branch 'backend_api_changes_for_classroom_admin' of https://github.com/Nik-09/oppia into classroom_admin_ui

Nik-09 committed:
Adds backend test for admin

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Fixes lint issues.

Nik-09 committed:
Adds frontend unit test

Nik-09 committed:
Adds classroom admin component tests.

Nik-09 committed:
Adds frontend unit tests

Nik-09 committed:
UI changes.

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Adds backend unit test coverage

Nik-09 committed:
Fixes typescript checks in strict mode.

Nik-09 committed:
Merge branch 'develop' into classroom_admin_ui

Nik-09 committed:
Rephrases test descriptions.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Merge branch 'classroom_admin_ui' of https://github.com/Nik-09/oppia into classroom_admin_ui

Nik-09 reviewed (commented):
Hi, @seanlip, can you please your feedback on this PR?
Thank you.

seanlip reviewed (changes requested):
Thanks @Nik-09! First, some general feedback:

- Please fill in the PR description overview at the top fully. You can use '#N/A' for point 1 if there's no issue number. But point 2 should be fully filled in for all PRs you open and before you request a review.
- With regards to the video, I have the following thoughts:
    - 0:06 "Generate classroom" doesn't really explain what's being generated, could you provide more detail about what exactly is being generated? See e.g. the examples before it.
    - You can use the green website navbar for the Classroom Admin page. It's basically similar in terms of framing to the Topic Editor page, for example.
    - Why does the classroom admin "existing classrooms" section take so long to load? It is too slow, I think.
    - Put the save/cancel buttons on the bottom right, not the bottom left. Put the save button on the far right.
    - After deleting the second classroom, it still shows up for quite a while before disappearing. Instead, it should be removed from the UI immediately.
    - What happens if you click "add new classroom" while a classroom editing form is open? Is it safer to disable that button (with a tooltip saying something like "Please save all edits before adding a new classroom") until editing has finished?
    - For validation, can you have the error message show up just below the entry field? See what we do on the signup/registration page when the user enters a duplicate username, for example.

Please let me know if you have any questions about the above. Thanks!

seanlip commented on core/controllers/admin.py:
```diff
@@ -673,6 +678,36 @@ def _generate_dummy_explorations(
+            classroom_dict = {
+                'classroom_id': new_classroom_id,
+                'name': new_classroom_name,
+                'url_fragment': '',
```
Would these fields even pass validation? I think it's fine to populate them if that is what you would expect from other classrooms.

Also, in general, you probably don't even want to create a classroom object at all in the backend unless the correct validations happen (e.g. url fragment is not empty).

Also, this looks like just an empty classroom. So I'm wondering what the point of having this functionality is. Why can't the admin just make a classroom through the standard approach? It does not feel like this saves them much work (unless it also generates the topics etc. within the classroom too).

Nik-09 replied:
Thank you for the clarification, I need to ask what would be better among the below two options.
(a). Provide 2 or 3 classroom options with a different number of topics and their dependencies. Like "Load Classroom A", "Load Classroom B", etc.
(b).  Provide a generic button using which an admin can create any number of classrooms with a fixed # topic and dependency graph.

I am asking this because both of the approaches have some benefits -- in (a) the admin can create 2 or 3 different classrooms with different topic dependencies, while in (b) the admin can create any # topic.

Please provide your suggestion in this context.

Thank you

seanlip replied:
I guess I'm wondering why you need this stuff at all. Couldn't the admin just make the classrooms manually if they wanted to? It doesn't seem that hard.

If you want to easily generate test data on the Activities tab for local testing, though, then I would have a button that's something like "Generate 2 classrooms, the first with 5 topics, the second with 1 topic; also generate all the topics and skills in those classrooms" and have the button actually generate the classrooms, topics and skills. It can be any reasonable sample data; the idea is to get a working setup that can be easily experimented with locally while reducing the cost to set that up. Note that this button would generate for a specific case, it doesn't need to take in any input fields etc. (Since the only users are developers they could probably make small adjustments to the code behind the button if needed.)

Nik-09 replied:
I have updated the section.
While generating the topics, I am confused about whether to generate the skill or not, because skills will not be required in the classroom admin feature. Only topics and their dependency suffice to make a graph.
Please provide your suggestion, on whether generating skills is essential or not.
Thank you.

seanlip replied:
Good question. After thinking about it, I would say, on balance -- yes, generate some skills for the topics. If the purpose is to make dummy data for simulation then it makes sense for this to have at least some of the stuff that would normally be expected, and skills are easy to delete if they're not wanted.

Nik-09 replied:
Thank you for the confirmation, I have added 1 skill/topic.

seanlip reviewed (changes requested):
For the PR, I took just a codeowners pass for now.

Nik-09 commented:
> 0:06 "Generate classroom" doesn't really explain what's being generated, could you provide more detail about what exactly is being generated? See e.g. the examples before it.

This point needs some clarification from my end. After discussion, I will update the instruction message here, until then I am keeping this point open.

> You can use the green website navbar for the Classroom Admin page. It's basically similar in terms of framing to the Topic Editor page, for example.

Updated
> Why does the classroom admin "existing classrooms" section take so long to load? It is too slow, I think.

I think this is because of my machine, I have updated the video in which the loading is bit fast, please take a look.
 
> Put the save/cancel buttons on the bottom right, not the bottom left. Put the save button on the far right.

Done

> After deleting the second classroom, it still shows up for quite a while before disappearing. Instead, it should be removed from the UI immediately.

(Same as of 3rd point.)

> What happens if you click "add new classroom" while a classroom editing form is open? Is it safer to disable that button (with a tooltip saying something like "Please save all edits before adding a new classroom") until editing has finished?

Thank you for the suggestion, I have updated the UI and video, please take a look.

> For validation, can you have the error message show up just below the entry field? See what we do on the signup/registration page when the user enters a duplicate username, for example.

Thank you, I have updated the UI and video, please take a look.

Nik-09 committed:
Addressed review comments

Nik-09 reviewed (commented):
Hi @seanlip, I have addressed the review comments, please take a look once you are free.

Thank you.

seanlip commented:
Re the video, it looks better with regards to the previous comments! Some follow-up thoughts:

- When editing an existing classroom, I would suggest graying out the "Add new classroom" button. It's currently still green, even though it's disabled.
- When you save edits to a classroom, all the fields remain editable and the cancel button remains clickable; this should probably not be the case. I would suggest making the save look instantaneous in the frontend and if the backend subsequently returns an error result, showing this in a blocking modal that forces reload of the page; otherwise, showing a green toast at the bottom right saying "Classroom details successfully edited."
- The "A classroom with this name already exists." message should disappear when the user starts making changes to the field again.
- The classroom URL fragment should be entered when creating a new classroom. It should also never be allowed to be empty at any point, similar to the classroom name. The URL fragment should only use lowercase alphabetic characters and hyphens.
- Could you show what happens if you edit a classroom after creating it, and set the name or url fragment to empty (or something invalid)? This should not be allowed and a save should not be possible.
- Change "Fill the basic details of a classroom" to "Add New Classroom".

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Removes dead code

Nik-09 committed:
Rephrases html text.

Nik-09 commented:
> When editing an existing classroom, I would suggest graying out the "Add new classroom" button. It's currently still green, even though it's disabled.

Done

> When you save edits to a classroom, all the fields remain editable and the cancel button remains clickable; this should probably not be the case. I would suggest making the save look instantaneous in the frontend and if the backend subsequently returns an error result, showing this in a blocking modal that forces reload of the page; otherwise, showing a green toast at the bottom right saying "Classroom details successfully edited."

I have added a check to disable the cancel button while saving the data, please take a look. For success toast, thank you for the suggestion I added while building but for some reason the toast was not working. Please take a look at the videos now if it looks good then I will leave or I have to dig a bit deeper and find the reason why the toast was not working (I think because of wrong imports while testing the toast).

> The "A classroom with this name already exists." message should disappear when the user starts making changes to the field again.

Updated

> The classroom URL fragment should be entered when creating a new classroom. It should also never be allowed to be empty at any point, similar to the classroom name. The URL fragment should only use lowercase alphabetic characters and hyphens.

Updated

> Could you show what happens if you edit a classroom after creating it, and set the name or url fragment to empty (or something invalid)? This should not be allowed and a save should not be possible.

Added in the video

> Change "Fill the basic details of a classroom" to "Add New Classroom".

Done

Nik-09 committed:
Fixes lint issues

seanlip commented:
Hi @Nik-09, this seems pretty good overall. I think the toast is optional, so don't worry about it.

The only things I noticed are:

- Some text errors: "**The** classroom name should not be empty." and similarly for the URL fragment.
- When you open the new classroom dialog, it's not great to see the error messages all at once. The user hasn't entered anything yet! Only show the error message if they enter something invalid. Also, for the URL fragment, show only one error message at a time.
- I suggest adding a "+" icon to the left of "Add new classroom".

Otherwise it looks great, thanks!

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Addressed review comments

Nik-09 commented:
> * Some text errors: "**The** classroom name should not be empty." and similarly for the URL fragment.

Updated

> * When you open the new classroom dialog, it's not great to see the error messages all at once. The user hasn't entered anything yet! Only show the error message if they enter something invalid. Also, for the URL fragment, show only one error message at a time.

I have updated the UI please take a look.

> * I suggest adding a "+" icon to the left of "Add new classroom".

Added

seanlip commented:
Hi @Nik-09, the first video is still wrong? But the second one looks good (though it doesn't cover the edit-in-place aspects of the first video). The only thing I saw in the second one was an issue with the error message: it should be "The classroom URL fragment should only contain lowercase letters separated by hyphens."

Nik-09 committed:
Updates html text.

Nik-09 commented:
> Hi @Nik-09, the first video is still wrong? But the second one looks good (though it doesn't cover the edit-in-place aspects of the first video). The only thing I saw in the second one was an issue with the error message: it should be "The classroom URL fragment should only contain lowercase letters separated by hyphens."

Updated the PR description with a single video, please take a look.

Nik-09 committed:
Rephrases test descriptions

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 reviewed (commented):
Hi, @seanlip I have updated the video and added 1 skill/topic for generating the dummy classroom, please take a look.

@kevintab95 @U8NWXD @gp201 please take a look at this PR and provide your suggestions once you are free.

Thank you

seanlip reviewed (approved):
@Nik-09 the UI looks great! Just one small nit: in the view on the main page (outside the modal), when you delete the classroom name entirely, the header of the accordion collapses. Don't do that -- instead, keep it the same height (and if it helps you can put something like "[New Classroom]" there if the field is empty).

Otherwise, LGTM and I have no other concerns (at least for the UI + codeowner files). Thanks!

U8NWXD commented on scripts/typescript_checks.py:
```diff
@@ -173,6 +173,7 @@
     'core/templates/pages/Base.ts',
     'core/templates/pages/admin-page/admin-page.import.ts',
     'core/templates/pages/blog-admin-page/blog-admin-page.import.ts',
+    'core/templates/pages/classroom-admin-page/classroom-admin-page.import.ts',
```
Can you strictly type this file instead? /cc @vojtechjelinek

vojtechjelinek replied:
Yeah, this should be typed.

Nik-09 replied:
Thank you @U8NWXD @vojtechjelinek for the confirmation.
I have updated the code, please take a look.

Nik-09 committed:
Addressed review comments

Nik-09 commented:
> @Nik-09 the UI looks great! Just one small nit: in the view on the main page (outside the modal), when you delete the classroom name entirely, the header of the accordion collapses. Don't do that -- instead, keep it the same height (and if it helps you can put something like "[New Classroom]" there if the field is empty).
> 
> Otherwise, LGTM and I have no other concerns (at least for the UI + codeowner files). Thanks!

@seanlip, I have updated the video, please take a look once you are free.
Thank you

seanlip commented:
LGTM, thanks @Nik-09. One further suggestion would be to maybe just gray the header text out a bit and italicize it when it's [New Classroom] so that it looks more like a placeholder, but this is minor and up to you. Otherwise I have nothing to add!

kevintab95 commented on core/domain/classroom_config_domain.py:
```diff
@@ -110,29 +111,69 @@ def to_dict(self) -> ClassroomDict:
+        if name == '':
+            raise utils.ValidationError('Name field should not be empty')
+
+        name_limit = constants.MAX_CHARS_IN_CLASSROOM_NAME
```
name_limit is confusing, why not just use constants.MAX_CHARS_IN_CLASSROOM_NAME inline?

Nik-09 replied:
Updated

kevintab95 commented on core/templates/domain/classroom/classroom-backend-api.service.ts:
```diff
@@ -250,6 +254,34 @@ export class ClassroomBackendApiService {
+      successCallback: (value: boolean) => void,
+      errorCallback: (errorResponse: HttpErrorResponse) => void
+  ): void {
+    let classroomUrlFragmentUrl = this.urlInterpolationService.interpolateUrl(
```
Prefer const since we are not modifying this.

Nik-09 replied:
Updated

kevintab95 commented on core/templates/domain/classroom/classroom-backend-api.service.ts:
```diff
@@ -250,6 +254,34 @@ export class ClassroomBackendApiService {
+    });
+  }
+
+  async doesClassroomWithUrlFragmentExist(classroomUrlFragment: string):
```
Should we append "Async" to this method name as suggested in the style guide [here](https://github.com/oppia/oppia/wiki/Coding-style-guide#javascript)?

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.html:
```diff
@@ -0,0 +1,265 @@
+                The classroom name should not be empty.
+              </p>
+              <p *ngIf="duplicateClassroomName" class="oppia-warning-text">
+                The classroom with this name already exists.
```
The -> A

Nik-09 replied:
Thank you, I have updated this part.

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.html:
```diff
@@ -0,0 +1,265 @@
+
+          <div class="row oppia-description-text-container">
+            <span class="col-lg-3 col-md-3 col-sm-3">
+              <strong>Url fragment</strong>
```
Url -> URL

kevintab95 replied:
Ditto below.

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -0,0 +1,346 @@
+  classroomDetailsIsShown: boolean = false;
+  classroomViewerMode: boolean = false;
+  classroomEditorMode: boolean = false;
+  savingClassroomData: boolean = false;
```
Maybe classroomDataSaveInProgress?

Nik-09 replied:
Thank you, I have updated this

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -0,0 +1,346 @@
+  classroomEditorMode: boolean = false;
+  savingClassroomData: boolean = false;
+
+  classroomNameExceedsMaxLen: boolean = false;
```
```suggestion
  classroomNameIsTooLong: boolean = false;
```

Ditto for URL fragment

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -0,0 +1,346 @@
+  classroomNameIsValid: boolean = true;
+
+  classroomUrlFragmentExceedsmaxLen: boolean = false;
+  emptyClassroomUrlFrgament: boolean = false;
```
Perhaps classroomUrlFragmentIsEmpty?

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -0,0 +1,346 @@
+
+  classroomUrlFragmentExceedsmaxLen: boolean = false;
+  emptyClassroomUrlFrgament: boolean = false;
+  duplicateClassroomUrlFragment: boolean = false;
```
Perhaps classroomUrlFragmentIsDuplicate? Otherwise this sounds like a function name.

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -0,0 +1,346 @@
+  classroomUrlFragmentExceedsmaxLen: boolean = false;
+  emptyClassroomUrlFrgament: boolean = false;
+  duplicateClassroomUrlFragment: boolean = false;
+  urlFragmentRegexMatched: boolean = true;
+  classroomUrlFragmentIsValid: boolean = true;
```
Why are these true by default?

Nik-09 replied:
When we are fetching the classroom data, then we are expecting to populate the valid data from the backend. Thus by default, I have set this as true. Please provide your view on this.
Thank you

kevintab95 replied:
In that case, while the data is being fetched, these variables don't represent the right information. We should make sure that all the variables represent the current state of things at all times.

Nik-09 replied:
updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -0,0 +1,346 @@
+
+        this.existingClassroomNames = (
+          Object.values(this.classroomIdToClassroomName));
+        let index = this.existingClassroomNames.indexOf(this.classroomName);
```
Prefer const.

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -0,0 +1,346 @@
+    if (this.classroomEditorMode) {
+      return;
+    }
+    this.classroomDetailsIsShown = true;
+    this.classroomViewerMode = true;
```
Should we wait for the Async call to complete before setting these?

Nik-09 replied:
Yes, we can enable the view mode after the completion of the Async call, the reason for the current initialization is –
- Each classroom tile in the card is clickable and the backend asynchronous call to fetch the classroom may take some time depending on various factors. So I am thinking that if in any case, the data fetching takes time then at least users know that the data is loading otherwise they may face a situation when they click the classroom tile and nothing happens for a moment.
- Please rectify my thought if I am thinking in the wrong direction.

Looking forward to your suggestion.
Thank you

kevintab95 replied:
Two things:
1. If we just need to identify if the async call is in proress, would it be sufficient to use a single flag? If, for some reason, we need several flags that need to be kept in sync, would it make sense to set/unset them in a method (avoids issues where future developers miss setting certain flags)?
2. Should there be any error handling in the case where the backend call fails (e.g. unset the flags / surface the error to the user)?

Nik-09 replied:
1. Updated
2. Added alert for error handling.

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -0,0 +1,346 @@
+  topicIdToPrerequisiteTopicIds: {[topicId: string]: string[]} = {};
+
+  pageIsInitialized: boolean = false;
+  classroomDataIsChanged: boolean = false;
```
What does this variable represent?

Nik-09 replied:
The variable is used to keep track of when the data in a classroom gets modified. When there is a change in classroom property this variable becomes true and due to this save button enables.

kevintab95 replied:
Consider a scenario where the value of the field has not been changed, but the users interacted with the field (e.g. edited it but changed it back to the original value or just clicked on the field and clicked outside). In that case, it seems like the variable may not be representing the correct information.

Would it make sense to use a common model across the different input fields (e.g. tmpClassroomData could be the main object, individual fields will refer to attributes of the object such as tmpClassroomData.urlFragment, tmpClassroomData.name etc. which will be passed to ngModel) and have the save button check if any of the attributes have changed and then enable/disable it accordingly?

Nik-09 replied:
Thank you @kevintab95 for the suggestion.
The updated code contains:
- A new model to track classroom fields.
- The blur method checks all the fields and then enables/disables the save button.

Please take a look, once you are free.

kevintab95 replied:
Do we still need `classroomDataIsChanged`?

Nik-09 replied:
I will say yes because on blur I have created a standard method that checks all the classroom fields and accordingly updates the variable. And this variable ultimately enables/disables the save button.

Please provide your view regarding this.

kevintab95 replied:
Sounds good, thanks!

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.html:
```diff
@@ -0,0 +1,82 @@
+                (ngModelChange)="onClassroomUrlFragmentChange()">
+      </textarea>
+      <p *ngIf="classroomUrlFragmentExceedsmaxLen" class="oppia-warning-text">
+        The classroom url fragment should contain at most 20 characters.
```
url -> URL

ditto below.

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.ts:
```diff
@@ -0,0 +1,174 @@
+  emptyClassroomUrlFrgament: boolean = false;
+  duplicateClassroomUrlFragment: boolean = false;
+  urlFragmentRegexMatched: boolean = true;
+  classroomUrlFragmentIsValid: boolean = false;
```
Given that the validations are similar between this component and the classroom admin page component, should we consider creating a service that handles the validation parts? It might help reduce some code duplication.

Nik-09 replied:
True, there is some code repetition my view regarding this is – 
In the classroom admin page component there are several properties apart from classroom name and URL fragment. So wouldn't it cause some confusion there because (after introducing a service) some properties are initialized and validated by an external service while some of the properties are initialized locally in the component?

Looking forward to your suggestion.
Thank you

kevintab95 replied:
I think we could still separate the common logic to a service, it may be that validation would be different in the two scenarios e.g. in the creation case, the validation may take into account the required URL and name fields whereas in the other component it will validate more fields.


I would recommend creating separate interfaces for a "new classroom" and an "existing classroom". A new classroom object can be constructed with a minimal set of properties whereas the existing classroom will be more exhaustive. Both the interfaces will contain a validate method, which will allow us to create separate classes with different validate  methods. We can define a type "ClassroomData" that will be the union of "NewClassroom" and "ExistingClassroom" -- this is the typescript way of implementing inheritance. The service will then deal only with ClassroomData. Calling the validate method of ClassroomData in the service will validate the relevant classroom object. Here's a sample implementation of what I mean:


```
interface NewClassroom {
    name: string;
    url: string;
    validate: () => boolean
}

interface ExistingClassroom extends NewClassroom {
    // The extra properties that an existing classroom may contain.
    extra1: number;
    extra2: number;
    validate: () => boolean
}

// Union type
type ClassroomData = NewClassroom | ExistingClassroom;

class NewClassroomData implements NewClassroom {
    name: string;
    url: string;
    constructor(name: string, url: string) {
        this.name = name;
        this.url = url;
    }
    validate(): boolean {
        return !!(this.name && this.url)
    }
}

class ExistingClassroomData implements ExistingClassroom {
    name: string;
    url: string;
    extra1: number;
    extra2: number;
    constructor(name: string, url: string, extra1: number, extra2: number) {
        this.name = name;
        this.url = url;
        this.extra1 = extra1;
        this.extra2 = extra2;
    }
    validate(): boolean {
        return !!(this.name && this.url && this.extra1 && this.extra2)
    }
}

const newClass = new NewClassroomData('new', '3');
const existingClass = new ExistingClassroomData('exist', 'url', 0, 1);


// In the service, we only use the type "ClassroomData". 
const validateClassroom = (classroom: ClassroomData) => {
    // Calling validate will do a dynamic dispatch and call the
    // validation method in the relevant classroom object.
    return classroom.validate();
}

// Outputs true
console.log(validateClassroom(newClass));

// Outputs false because extra1 is 0
console.log(validateClassroom(existingClass));
```

Nik-09 replied:
Thank you @kevintab95 this code explanation really helped a lot.

I have created a new file "classroom-admin.model" which contains the classroom-related fields, variables, and methods.
The model also helped me to reduce duplicate code. 
Please take a look at the updated code, once you are free.

Thank you

kevintab95 replied:
Thanks @Nik-09. I think this is heading in the right direction.

I suggested creating 2 models for the two kind of classroom data that exist i.e. creation of two interfaces with the specific properties it is validating and using dynamic dispatch to call the relevant validation method. Should we do that? This is a better approach than manually suppressing warnings in the main model (which is confusing and prone to errors when not used correctly.)

Nik-09 replied:
Updated by doing the following tasks
- Split the class into two (ExistingClassroomData and NewClassroomData)
- Created a new service

Please take a look.

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.ts:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  ngOnInit(): void {
+    this.getNewClassroomId();
```
Should we only allow user interaction with the component once the classroom ID is set?

Nik-09 replied:
No, basically here I am trying to get the classroom ID once the component gets initialized. I have done this in order to save some time. Please provide your suggestion in this context.
Thank you

kevintab95 replied:
Note that the `getNewClassroomId` actually triggers an async call. In slow networks, there is a possibility of a race condition where the newClassroomId is set late, in the meantime we should ensure that the UI behaves correctly -- user should be able to interact with it only when the UI is ready e.g. edit/save buttons are disabled until the required variables for the component are set correctly.

Nik-09 replied:
Thank you for the explanation, I have updated the code, please take a look.

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.ts:
```diff
@@ -0,0 +1,174 @@
+  newClassroomName: string = '';
+  newClassroomId: string = '';
+  newClassroomUrlFragment: string = '';
+  creatingNewClassroom: boolean = false;
```
This sounds like a function name (it starts with a verb.) Should we rename it?

Nik-09 replied:
Updated

kevintab95 reviewed (changes requested):
Thanks @Nik-09! Left some comments. PTAL

gp201 commented on core/templates/pages/admin-page/activities-tab/admin-dev-mode-activities-tab.component.spec.ts:
```diff
@@ -396,6 +396,41 @@ describe('Admin dev mode activities tab', () => {
+      expect(component.setStatusMessage.emit)
+        .toHaveBeenCalledWith('Processing...');
+
+      fixture.whenStable().then(() => {
```
Please add a comment explaining why you used `fixture.whenStable()` here.

Nik-09 replied:
Added a comment

gp201 commented on core/templates/pages/admin-page/activities-tab/admin-dev-mode-activities-tab.component.spec.ts:
```diff
@@ -396,6 +396,41 @@ describe('Admin dev mode activities tab', () => {
+        expect(component.setStatusMessage.emit)
+          .toHaveBeenCalledWith('Processing...');
+
+        fixture.whenStable().then(() => {
```
Ditto

Nik-09 replied:
Added

gp201 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.spec.ts:
```diff
@@ -0,0 +1,300 @@
+      expect(componentInstance.classroomNameIsValid).toBeFalse();
+      expect(componentInstance.duplicateClassroomName).toBeFalse();
+
+      componentInstance.existingClassroomNames = ['physics', 'chemistry'];
+
+      componentInstance.newClassroomName = 'physics';
```
```suggestion
      componentInstance.existingClassroomNames = ['physics', 'chemistry'];
      componentInstance.newClassroomName = 'physics';
```

gp201 replied:
same everywhere else

Nik-09 replied:
Updated

gp201 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.spec.ts:
```diff
@@ -0,0 +1,300 @@
+      expect(componentInstance.classroomUrlFragmentIsValid).toBeFalse();
+      expect(componentInstance.urlFragmentRegexMatched).toBeTrue();
+
+      componentInstance.newClassroomUrlFragment = 'Incorrect-url';
+      componentInstance.onClassroomUrlFragmentChange();
```
```suggestion
      componentInstance.newClassroomUrlFragment = 'Incorrect-url';

      componentInstance.onClassroomUrlFragmentChange();
```

Nik-09 replied:
Updated

gp201 reviewed (changes requested):
@Nik-09 PTAL. left a couple of comments

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Fixes lint issues

Nik-09 reviewed (commented):
Hi, @kevintab95 @gp201 I have addressed the review comments.

Please take a look at this PR, once you are free.
Thank you

Nik-09 committed:
Fixes merge conflict

Nik-09 reviewed (commented):
Hi, @kevintab95 @gp201 I have updated the PR as per the review comments, please take a look once you are free.

Thank you.

gp201 reviewed (approved):
LGTM

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.html:
```diff
@@ -0,0 +1,82 @@
+  .new-classroom-input-data {
+    margin: 10px;
+  }
+  :host /deep/ .oppia-create-classroom-spinning-button circle {
```
Why do we need to use `/deep/` here? It breaks view encapsulation and must be avoided.

Nik-09 replied:
The progress spinner (mat-spinner), contains the stroke color `#00645c` which is declared in the 'oppia-material.css'. But in this context, I wanted a stroke color of `#ffffff`. 
So in this view, the spinner always presents `#00645c`. After trying several solutions I came up with this (https://stackoverflow.com/questions/11509830/how-to-add-color-to-githubs-readme-md-file).

Also, in some pages of Oppia, 
- state-skill-editor
- blog-dashboard-page

we are using /deep/ in the CSS.

Thus I have used it here. Looking forward to your suggestions.

kevintab95 replied:
Thanks for checking. In general, we should not be allowing more use of /deep/. It is deprecated and support will be dropped, from what I can tell. Can we at least add a comment explaining why this is required here?

Also, should we use `:ng-deep` instead for wider compatibility as recommended [here](https://angular.io/guide/component-styles#:~:text=Until%20then%20%3A%3Ang%2Ddeep%20should%20be%20preferred%20for%20a%20broader%20compatibility%20with%20the%20tools.)?

Nik-09 replied:
Thank you Kevin for the suggestion, I have tried the solution but it is not working and I am not sure about the reason. Please take a look.

https://user-images.githubusercontent.com/57531197/192355545-5fc72876-8510-4f18-ab7f-3cdb10b1f4d4.mp4

Nik-09 replied:
Thank you Kevin for the suggestion, I have tried the solution but it is not working and I am not sure about the reason. Please take a look.

https://user-images.githubusercontent.com/57531197/192355545-5fc72876-8510-4f18-ab7f-3cdb10b1f4d4.mp4

kevintab95 replied:
I would suggest creating a TODO for this given that (1) we are short on time and (2) there are existing instances of `/deep/` in the codebase. We will need to handle this in the future though when Angular decides to stop support for shadow DOM piercing CSS.

Nik-09 replied:
Added a todo comment, please take a look.

kevintab95 reviewed (changes requested):
Thanks @Nik-09! I've left some comments. PTAL

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Removes duplicate naming for dummy topics

Nik-09 committed:
Refactored classroom component as per review comments

Nik-09 committed:
Adds frontend unit test.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Fixes lint issues

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.ts:
```diff
@@ -0,0 +1,90 @@
+    this.newClassroomCreationInProgress = true;
+    this.classroomUrlFragmentIsDuplicate = false;
+
+    this.classroomBackendApiService.doesClassroomWithUrlFragmentExistAsync(
+      this.newClassroom.urlFragment).then(response => {
+      if (response) {
+        this.classroomUrlFragmentIsDuplicate = true;
+        this.newClassroom.classroomUrlFragmentIsValid = false;
+        this.newClassroomCreationInProgress = false;
+        return;
+      }
```
Should we do this validation as part of updating the URL fragment?

Nik-09 replied:
Thank you, I have updated this.

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.ts:
```diff
@@ -0,0 +1,90 @@
+        return;
+      }
+
+      this.classroomBackendApiService.getNewClassroomIdAsync().then(
+        classroomId => {
```
Have you considered creating a classroom ID on opening the modal instead?

Nik-09 replied:
I have thought about this and not implemented it because -- 
It may be possible that any user only to check the functionality clicks the button to create a new classroom without having the intention to create one. If clicking the button is associated with a backend call to get a classroom ID, and the classroom is not created then maybe it is an expensive call. Thus I have implemented it in such a way that while creating the classroom the new classroom ID is fetched from the backend. 

Please provide your view on this,

kevintab95 reviewed (changes requested):
Thanks @Nik-09! This looks a lot better. Left some comments. PTAL

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Fixes lint issues

Nik-09 reviewed (commented):
Hi, @kevintab95 I have addressed all of your review comments, please take a look.

I have not written the unit tests for the newly added service file because I wanted to take the review on it first.

Thank you

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.ts:
```diff
@@ -0,0 +1,85 @@
+
+  ngOnInit(): void {
+    this.newClassroom = new NewClassroomData('', '', '');
+    this.newClassroom.classroomNameIsValid = false;
+    this.newClassroom.classroomUrlFragmentIsValid = false;
```
Please avoid accessing class properties in this manner. The properties should be private and accessed using relevant getters/setters.

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.ts:
```diff
@@ -0,0 +1,85 @@
+
+    // The error messages for classroom name and URL fragment, should not be
+    // presented to the user when the modal is just opened.
+    this.classroomAdminDataService.supressClassroomNameErrorMessages(
+      this.newClassroom);
+    this.classroomAdminDataService.supressClassroomUrlFragmentErrorMessages(
+      this.newClassroom);
```
Why not just have a validate method defined in the classroom models that sets the relevant flags depending on its type? Rather than having to deal with this manual book-keeping (i.e. in this scenario, everywhere a developer uses "newClassroom", they will need to diligently suppress warnings manually -- this can be error prone)

Nik-09 replied:
Added two different validations in two models (1 for name and other for URL fragment)

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin.model.ts:
```diff
@@ -0,0 +1,144 @@
+}
+
+
+export class ExistingClassroomData implements ExistingClassroom {
```
Please only have one class in a model TS file.

Nik-09 replied:
Seperated

kevintab95 reviewed (changes requested):
Thanks @Nik-09! I think this is getting closer! Left a couple of comments. PTAL

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Addressed review comments.

kevintab95 commented on core/templates/pages/classroom-admin-page/existing-classroom-admin.model.ts:
```diff
@@ -0,0 +1,204 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Existing classroom model.
```
Should we remove "admin" part from the filename (ditto for the other new classroom model)?

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/existing-classroom-admin.model.ts:
```diff
@@ -0,0 +1,204 @@
+  getTopicListIntro: () => string;
+  getTopicIdToPrerequisiteTopicId: () => TopicIdToPrerequisiteTopicIds;
+  IsClassroomNameValid: () => ValidateClassroomFieldResponse;
+  IsClassroomUrlFragmentIsValid: () => ValidateClassroomFieldResponse;
```
Please start function names with a lowercase letter.

Nik-09 replied:
Removed this function.

kevintab95 commented on core/templates/pages/classroom-admin-page/existing-classroom-admin.model.ts:
```diff
@@ -0,0 +1,204 @@
+import { NewClassroom } from './new-classroom-admin.model';
+
+
+export interface ValidateClassroomFieldResponse {
```
Avoid starting with a verb unless it is a function.

Nik-09 replied:
Removed

kevintab95 commented on core/templates/pages/classroom-admin-page/existing-classroom-admin.model.ts:
```diff
@@ -0,0 +1,204 @@
+
+export interface ValidateClassroomFieldResponse {
+  result: boolean;
+  type: string;
```
Unclear what type means in this context.

kevintab95 replied:
Do we need a custom type here? Would it be sufficient to have the validation method to be the equivalent of "getValidationErrors" which just returns a list of error strings. If there are no errors, the list is empty.

Nik-09 replied:
Removed.

kevintab95 reviewed (changes requested):
Thanks @Nik-09! I think this looks a lot better, great work! I've left a few minor comments to hopefully simplify things further. PTAL.

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Adds tests for new classroom model

Nik-09 reviewed (commented):
Working on the tests, and I will update all the tests in the next commit.
Thank you

Nik-09 reviewed (commented):
Thank you @kevintab95 for the review, I have updated the code, please take a look once you are free.

Hi, @srijanreddy98 please take a look at this PR, once you are free.

Thank you

kevintab95 commented on core/templates/pages/classroom-admin-page/existing-classroom.model.ts:
```diff
@@ -0,0 +1,179 @@
+    this._topicListIntro = topicListIntro;
+  }
+
+  getClassroomNamValidationError(): string {
```
Typo: Nam -> Name

Should we rename this to `getClassroomNameValidationErrors` instead?

Nik-09 replied:
Done

kevintab95 commented on core/templates/pages/classroom-admin-page/existing-classroom.model.ts:
```diff
@@ -0,0 +1,179 @@
+    return errorMsg;
+  }
+
+  getClassroomUrlValidationError(): string {
```
```suggestion
  getClassroomUrlValidationErrors(): string {
```

Nik-09 replied:
Done

kevintab95 commented on core/templates/pages/classroom-admin-page/new-classroom.model.ts:
```diff
@@ -0,0 +1,119 @@
+    this._urlFragment = urlFragment;
+  }
+
+  getClassroomNamValidationError(): string {
```
Ditto regarding naming.

kevintab95 replied:
Also, are these checks the same in the two models? Should we combine / reuse it to reduce duplicate code?

Nik-09 replied:
Done

kevintab95 commented on core/templates/pages/classroom-admin-page/new-classroom.model.ts:
```diff
@@ -0,0 +1,119 @@
+    return errorMsg;
+  }
+
+  getClassroomUrlValidationError(): string {
```
Ditto: naming

Nik-09 replied:
Done

kevintab95 reviewed (changes requested):
Thanks @Nik-09! This is great, I just have few nits remaining. PTAL

Nik-09 committed:
Fixes frontend tests

Nik-09 reviewed (commented):
Hi, @kevintab95 I have addressed your review comments please take a look.
Thank you.

kevintab95 reviewed (approved):
LGTM, thanks @Nik-09!

srijanreddy98 reviewed (approved):
LGTM!
